consensus/miner_signalled_consensus.md

```diff
@@ -0,0 +1,200 @@
+be guaranteed that the best valid chain were the longest.  So such
+global decision would be ineffective in preventing chain forks with
+distinct consensus protocol versions from being stored locally for the
+same height.  As a result, the user experience would be worse:
+- By taking the decision globally, on bootstrap a node may initially follow a
+  (valid) chain fork distinct from the best valid one hence take the wrong
+  decision.
+- As the best valid chain is not necessarily the longest one, by taking the
+  decision globally the node may anyway taint the storage with blocks with
+  eventually wrong consensus protocol.
+
+### Value of the info field not necessarily part of new consensus protocol
+
+The miner signalling is orthogonal to the validation rules of the new
+consensus protocol for the sake of simplicity of the implementations.
+
+This is not detrimental as the new consensus protocol may include
+arbitrary constraints.
+
+## Backwards Compatibility
+
+The signalling using the block header field `info` is backward compatible, as
+the meaning of the value of such field is not under consensus.
+
+The activation of the new consensus protocol based on miner signalling is meant
+to be a temporary measure during the time interval when the new protocol may be
+activated on the network while the node is running. Once the network reasonably
+settles on whether to activate the new consensus protocol, the user:
+- Is meant to enforce such decision on the eventual node restart either by
+  configuration or by upgrading the node to a new version.
+- Shall be recommended to discard the persisted storage on the eventual node
+  restart - as distinct persisted chain forks may have distinct consensus
+  protocols at the same height.
+
+In order to enable non-mutually authorized channel transactions to be
+included on chain with mutually-authorized-agreed-offline payload,
+channel transactions valid under the current consensus protocol must
+be valid also under the proposed new consensus protocol.
+
+## (possible) Future extensions
+
+### Simultaneous consensus proposals
+
+The usage of the block header field `info` offers room for optimization for
+accommodating simultaneous hard-forking or soft-forking consensus proposals, by
+interpreting such `info` field as a bitmask rather than a value.
+
+### Soft-forking consensus proposals
+
+This proposal may be extended to soft-forking consensus proposals - as opposed
+to only hard-forking ones - by relaxing the strict monotonicity of consensus
+versions.
+
+Such extension should preferably be accompanied by a working implementation.
+
+## Implementation
+
+Any implementation needs to cater for the performance of the aggregation of the
+signalling e.g. slow multiple lookups on the storage.
+
+Any implementation needs to consider that the storage may contain blocks with
+the old and new consensus protocols at heights at which the new consensus
```

Comment:
The node has _one_ chain at any point in time, there will not be multiple key-blocks at the same height, I don't understand this?!

Reply:
```
                         [3] --- [4] --- [5] --- [6] --- [7] --- [8] ---  <- High difficulty
                          |
 [0] --- [1] --- [2] --- [3] --- [4] --- [5] --- [6] --- [7] --- [8] ---  <- Low difficulty
                  H1      HS            HE - 1    HE    H2 - 1    H2
```
There is just low difficulty chain at first and the fork signalling result is true (so switching to the new protocol at H2). Then there is a chain fork at height HS, this chain has higher difficulty, so it wins. The fork signalling result for high difficulty chain is false (so not switching to the new protocol at H2). Now there are 2 blocks in the db with different protocol versions at height H2 (but only the one in high difficulty chain is "active").